p4-16/psa/psa.p4

```diff
@@ -83,15 +135,64 @@ const   CloneSessionId_t PSA_CLONE_SESSION_TO_CPU = unspecified;
+typedef bit<8>  ClassOfServiceInHeaderUint_t;
+typedef bit<16> PacketLengthInHeaderUint_t;
+typedef bit<16> EgressInstanceInHeaderUint_t;
+typedef bit<64> TimestampInHeaderUint_t;
 // END:Type_defns2
 
+/* The INT_TO_HEADER macros exist for use in those cases when one has
+ * a value of type <name>_t (an value INTernal to the data path) and
+ * you wish to assign it to a field of type <name>InHeader_t inside a
+ * header that will be sent to the CPU port.
+ *
+ * The HEADER_TO_INT macros exist for assigning in the opposite
+ * direction, typically for assigning a value in a header received
+ * from the CPU port, to a value you wish to use in the rest of your
+ * code, which uses less storage space.
+ *
+ * The reason that three casts are needed is that each of the original
+ * and target types is declared via P4_16 'type', so without a cast
+ * they can only be assigned to values of that identical type.  The
+ * first cast changes it from the original 'type' to a 'bit<W1>' value
+ * of the same bit width W1.  The second cast changes its bit width,
+ * either prepending 0s if it becomes wider, or discarding the most
+ * significant bits if it becomes narrower.  The third cast changes it
+ * from a 'bit<W2>' value to the final 'type', with the same width
+ * W2. */
+
+#define PSA_PORT_ID_HEADER_TO_INT(x) ((PortId_t) (PortIdUint_t) (PortIdInHeaderUint_t) (x))
+#define PSA_MULTICAST_GROUP_HEADER_TO_INT(x) ((MulticastGroup_t) (MulticastGroupUint_t) (MulticastGroupInHeaderUint_t) (x))
+#define PSA_CLONE_SESSION_HEADER_TO_INT(x) ((CloneSessionId_t) (CloneSessionIdUint_t) (CloneSessionIdInHeaderUint_t) (x))
+#define PSA_CLASS_OF_SERVICE_HEADER_TO_INT(x) ((ClassOfService_t) (ClassOfServiceUint_t) (ClassOfServiceInHeaderUint_t) (x))
+#define PSA_PACKET_LENGTH_HEADER_TO_INT(x) ((PacketLength_t) (PacketLengthUint_t) (PacketLengthInHeaderUint_t) (x))
+#define PSA_EGRESS_INSTANCE_HEADER_TO_INT(x) ((EgressInstance_t) (EgressInstanceUint_t) (EgressInstanceInHeaderUint_t) (x))
+#define PSA_TIMESTAMP_HEADER_TO_INT(x) ((Timestamp_t) (TimestampUint_t) (TimestampInHeaderUint_t) (x))
+
+#define PSA_PORT_ID_INT_TO_HEADER(x) ((PortIdInHeader_t) ((PortIdInHeaderUint_t) ((PortIdUint_t) (x))))
+#define PSA_MULTICAST_GROUP_INT_TO_HEADER(x) ((MulticastGroupInHeader_t) (MulticastGroupInHeaderUint_t) (MulticastGroupUint_t) (x))
+#define PSA_CLONE_SESSION_INT_TO_HEADER(x) ((CloneSessionIdInHeader_t) (CloneSessionIdInHeaderUint_t) (CloneSessionIdUint_t) (x))
+#define PSA_CLASS_OF_SERVICE_INT_TO_HEADER(x) ((ClassOfServiceInHeader_t) (ClassOfServiceInHeaderUint_t) (ClassOfServiceUint_t) (x))
+#define PSA_PACKET_LENGTH_INT_TO_HEADER(x) ((PacketLengthInHeader_t) (PacketLengthInHeaderUint_t) (PacketLengthUint_t) (x))
+#define PSA_EGRESS_INSTANCE_INT_TO_HEADER(x) ((EgressInstanceInHeader_t) (EgressInstanceInHeaderUint_t) (EgressInstanceUint_t) (x))
+#define PSA_TIMESTAMP_INT_TO_HEADER(x) ((TimestampInHeader_t) (TimestampInHeaderUint_t) (TimestampUint_t) (x))
```

Comment:
Seriously? Is this the best we can do? I usually find that when we need to do this kind of stuff, something is wrong with the design.

Reply:
There may be something wrong with the design.  No argument there.  Frankly, right now I don't have any ideas to make it better, unless we consider it to be a better design if we force PSA program writers to add annotations to _every_ occurrence of values of these types requiring translation, and they get a buggy system if they forget.

These macros are only useful when dealing with headers in data-plane-to-controller and controller-to-data-plane packets, if that helps you swallow the nasty-tasting-pill any more easily.

Reply:
I don't know if this helps with thinking about the problem, but if we want the compiler to be able to determine that a value of type `PortId_t` is actually `PortId_t`, and not some other type, imagine we tried to do it with typedefs, which allow arithmetic on them with other things that are the same bit width.

```
typedef bit<9> PortId_t;
typedef bit<9> ClassOfService_t;

// ... later in the program ...

PortId_t x;
ClassOfService_t y;

    table foo {
        key = {
            x + y : exact;
        }
        // ... rest of table defn here ...
    }
```

What is the type of that table key?  If you say PortId_t, why shouldn't it be ClassOfService_t?  What is a deterministic rule you are happy living with that all P4 compilers can use to find the answer?

And a program can have any number of typedefs with the same bit width, not just 2.

Reply:
I know @jnfoster believes there is a better way of approaching this whole area, but may not have free time to work on it.  I'm not trying to twist your arm on this, Nate, if that is still the case, but if you might be able to find 30 mins to brainstorm any ideas you have had on the topic with Calin and I, maybe we could take the ball and try running with it.

Reply:
One of Calin or Nate proposed the idea that became this PR, which would reduce the number of casts needed from 3 down to 1: https://github.com/p4lang/p4-spec/issues/682.  My only plan for a change in this area in the next commit is to replace the macro definitions with P4_16 functions that do the same thing, since macros with arguments are not required by the P4_16 language spec.